scapy/contrib/isotp.py

```diff
@@ -1611,3 +1613,391 @@ def recv(self, x=0xffff):
+# #################### ISOTPSCAN ####################################
+# ###################################################################
+def send_multiple(sock, packet, scan_range, number_of_packets):
+    """ Send multiple packets at once
+
+    Args:
+        sock: socket for can interface
+        packet: packet to send
+        scan_range: id range which will be scanned
+        number_of_packets: number of packets send
+    """
+    if scan_range[-1] > 0x7FF:
+        if scan_range[0] > 0x7FF:
+            print("Invalid ID range.")
+        else:
+            min_id = scan_range[0]
+            for i in range(0, ((0x7FF - scan_range[0]) + 1)):
+                scan_range[i] = min_id
+                min_id = min_id + 1
+
+    for i in list(scan_range)[::(number_of_packets + 1)]:
+        for j in range(i, (i + number_of_packets + 1)):
+            packet.identifier = j
+            sock.send(packet)
+
+
+def send_multiple_ext(sock, ext_id, packet, number_of_packets):
+    """ Send multiple packets with extended addresses at once
+
+    Args:
+        sock: socket for can interface
+        ext_id: extended id. First id to send.
+        packet: packet to send
+        number_of_packets: number of packets send
+
+    This function is used for scanning with extended addresses.
+    It sends multiple packets at once. The number of packets
+    is defined in the number_of_packets variable.
+    It only iterates the extended ID, NOT the actual ID of the packet.
+    This method is used in extended scan function.
+    """
+    end_id = min(ext_id + number_of_packets, 255)
+    for i in range(ext_id, end_id + 1):
+        packet.extended_address = i
+        sock.send(packet)
+
+
+def get_isotp_packet(identifier=0x0, extended=False):
+    """ Craft ISO TP packet
+    Args:
+        identifier: identifier of crafted packet
+        extended: boolean if packet uses extended address
+    """
+
+    if extended:
+        pkt = ISOTPHeaderEA() / ISOTP_FF()
+        pkt.extended_address = 0
+        pkt.data = b'\x00\x00\x00\x00\x00'
+    else:
+        pkt = ISOTPHeader() / ISOTP_FF()
+        pkt.data = b'\x00\x00\x00\x00\x00\x00'
+
+    pkt.identifier = identifier
+    pkt.message_size = 100
+    return pkt
+
+
+def filter_periodic_packets(packet_list, verbose=False):
+    """ Filter for periodic packets
+
+    Args:
+        packet_list: Dictionary with Send-to-ID as key and a tuple
+        (received packet, Recv_ID)
+        verbose: Displays further information
+
+    ISOTP-Filter for periodic packets (same ID, always same timegap)
+    Deletes periodic packets in packet_list
+    """
+
+    delete_list = set()
+    filter_dict = {}
+
+    for _, value in packet_list.items():
+        pkt = value[0]
+        idn = value[1]
+        if idn not in filter_dict.keys():
+            filter_dict[idn] = [pkt]
+        else:
+            filter_dict[idn] += [pkt]
+
+    for idn in filter_dict:
```

Comment:
`for key, value in filter_dict:`

Reply:
python2 and python3 have different behaviours

Reply:
You could use `six.iteritems(filter_dict)`